core/broadcast.go

```diff
@@ -270,46 +267,73 @@ func (d *dispatcher) stop() {
 }
 
 type sender struct {
-	ctx    context.Context
-	l      log.Logger
-	client net.ProtocolClient
-	to     net.Peer
-	newCh  chan broadcastPacket
+	ctx      context.Context
+	l        log.Logger
+	client   net.ProtocolClient
+	to       net.Peer
+	directCh chan broadcastPacket
+	newCh    chan broadcastPacket
 }
 
 func newSender(ctx context.Context, l log.Logger, client net.ProtocolClient, to net.Peer, queueSize int) *sender {
 	return &sender{
-		ctx:    ctx,
-		l:      l,
-		client: client,
-		to:     to,
-		newCh:  make(chan broadcastPacket, queueSize),
+		ctx:      ctx,
+		l:        l,
+		client:   client,
+		to:       to,
+		directCh: make(chan broadcastPacket, 1),
+		newCh:    make(chan broadcastPacket, queueSize),
 	}
 }
 
-func (s *sender) sendPacket(p broadcastPacket) {
-	select {
-	case s.newCh <- p:
-	default:
-		s.l.Debug("broadcast", "sender queue full", "endpoint", s.to.Address())
+func (s *sender) sendPacket(p broadcastPacket, direct bool) {
+	if direct {
+		s.directCh <- p
+	} else {
+		select {
+		case s.newCh <- p:
+		default:
+			s.l.Debug("broadcast", "sender queue full", "endpoint", s.to.Address())
+		}
 	}
 }
 
 func (s *sender) run() {
-	for newPacket := range s.newCh {
-		s.sendDirect(newPacket)
+	directOpen := true
+	normalOpen := true
+	for directOpen || normalOpen {
+		for directOpen {
+			select {
+			case p, ok := <-s.directCh:
+				if !ok {
+					directOpen = false
+					continue
+				}
+				s.sendDirect(p)
+			default:
+				break
```

Comment:
Moreover, this "default:break" means that when I'm not sending anything **yet**, this default break will always kick in and this will be a continuous loop until there is a message to send...